baselines/driving_smarts/benchmark.py

```diff
@@ -0,0 +1,140 @@
+        scenario=env_config["scenario"],
+        **env_config["shared_params"],
+    )
+    env = EpisodeLimit(env, episodes)
+    env = CompetitionMetrics(env)
+    agent = agent_registry.make_agent(
+        locator=agent_config["locator"],
+        **agent_config["params"],
+    )
+
+    observation, info = env.reset()
+    try:
+        while not info.get("reached_episode_limit"):
+            try:
+                # action: [global x-coordinate, global y-coordinate]
+                action = {
+                    agent_id: agent.act(obs) for agent_id, obs in observation.items()
+                }
+                # assert env.action_space.contains(action)
+            except Exception:
+                logging.error("Policy robustness failed.")
+                # env.mark_policy_failure()
+                terminated, truncated = True, True
+            else:
+                observation, reward, terminated, truncated, info = env.step(action)
+            if terminated or truncated:
+                observation, info = env.reset()
+    finally:
+        score = env.score()
+        env.close()
+    return name, score
+
+
+def task_iterator(env_configs, benchmark_config):
+    ray.init(num_cpus=4, log_to_driver=LOG_WORKERS)
+    try:
+        max_queued_tasks = 20
+        unfinished_refs = []
+        for name, env_config in env_configs.items():
+            if len(unfinished_refs) >= max_queued_tasks:
+                ready_refs, unfinished_refs = ray.wait(unfinished_refs, num_returns=1)
+                for name, score in ray.get(ready_refs):
+                    # sum scores
+                    yield name, score
+            print(f"Evaluating {env_config['scenario']}...")
+            unfinished_refs.append(
+                _eval_worker.remote(
+                    name=name,
+                    env_config=env_config,
+                    episodes=benchmark_config["eval_episodes"],
+                    agent_config=benchmark_config["agent"],
+                )
+            )
+        for name, score in ray.get(unfinished_refs):
+            yield name, score
+    finally:
+        ray.shutdown()
+
+
+def benchmark(benchmark_config):
+    print(f"Starting `{benchmark_config['name']}` benchmark.")
+    env_configs = {}
+    for scenario in benchmark_config["standard_env"]["scenarios"]:
+        env_configs[f"{scenario}"] = dict(
+            env=benchmark_config["standard_env"]["env"],
+            scenario=scenario,
+            shared_params=benchmark_config["shared_env_params"],
+        )
+    ## TODO MTA bubble env
+    # for seed in config["bubble_env"]["scenarios"]:
+    #     env_configs[f"bubble_env_{seed}"] = partial(
+    #         env=config["bubble_env"]["env"],
+    #         scenario=seed,
+    #         shared_params=config["shared_env_params"],
+    #     )
+    # TODO MTA: naturalistic
+    total_score = {}
+
+    def sum_scores(s, o):
+        return {k: s.get(k, 0) + o.get(k, 0) for k in set(o)}
```

Comment:
Why do we use a default value of `0` and what is the impact on the score's fairness?

Reply:
Should all keys already exist by the time this is run?

Reply:
+ `env.score()` should give values for all keys.
+ so all keys should be available in `o`.
+ keys will be missing in `s` in the very first call since `total_score={}`.
+ to avoid this, we can define a namedtuple for score in `metrics.py`.
+ then initiate total_score = empty_score_namedtuple

Reply:
Right. Or I could record all of them and sum them all at the end. I will just add a breakdown.

Reply:
I am seeing negative completion sometimes. I assume that is because the agent is going backwards?

Reply:
We should check why it is happening. Ideally, `0 <= Score.completion <= 1`.